File: leetcode/0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.js
function diffWaysToCompute(expression) {
    const result = [];

    // 수식의 각 문자 확인
    for (let i = 0; i < expression.length; i++) {
        const operator = expression[i];

        // 연산자를 발견한 경우
        if (
            operator === '+' ||
            operator === '-' ||
            operator === '*'
        ) {
            // 왼쪽 부분 계산
            const left = diffWaysToCompute(
                expression.substring(0, i)
            );

            // 오른쪽 부분 계산
            const right = diffWaysToCompute(
                expression.substring(i + 1)
            );

            // 모든 가능한 결과 조합
            for (const l of left) {
                for (const r of right) {
                    if (operator === '+') {
                        result.push(l + r);
                    } else if (operator === '-') {
                        result.push(l - r);
                    } else {
                        result.push(l * r);
                    }
                }
            }
        }
    }

    // 연산자가 없는 경우 숫자 반환
    if (result.length === 0) {
        return [Number(expression)];
    }

    return result;
}

File: 0241-different-ways-to-add-parentheses/0241-different-ways-to-add-parentheses.java
import java.util.*;

class Solution {

    public List<Integer> diffWaysToCompute(String expression) {
        List<Integer> result = new ArrayList<>();

        // 연산자가 없는 경우 숫자 반환
        boolean hasOperator = false;

        for (int i = 0; i < expression.length(); i++) {
            char operator = expression.charAt(i);

            if (operator == '+' ||
                operator == '-' ||
                operator == '*') {

                hasOperator = true;

                // 왼쪽 부분 계산
                List<Integer> left =
                    diffWaysToCompute(expression.substring(0, i));

                // 오른쪽 부분 계산
                List<Integer> right =
                    diffWaysToCompute(expression.substring(i + 1));

                // 모든 가능한 조합 계산
                for (int l : left) {
                    for (int r : right) {
                        if (operator == '+') {
                            result.add(l + r);
                        } else if (operator == '-') {
                            result.add(l - r);
                        } else {
                            result.add(l * r);
                        }
                    }
                }
            }
        }

        // 연산자가 없는 경우 숫자
        if (!hasOperator) {
            result.add(Integer.parseInt(expression));
        }

        return result;
    }
}